lib/datadog/tracing/contrib/utils/quantization/http.rb

```diff
@@ -14,20 +14,25 @@ module HTTP
 
             PLACEHOLDER = '?'.freeze
 
+            # taken from Ruby 3.1.2 lib/uri/rfc3986_parser.rb and adjusted to stop at path
+            RFC3986_URL_BASE = /\A(?<URI>(?<scheme>[A-Za-z][+\-.0-9A-Za-z]*):(?<hier-part>\/\/(?<authority>(?:(?<userinfo>(?:%\h\h|[!$&-.0-;=A-Z_a-z~])*)@)?(?<host>(?<IP-literal>\[(?:(?<IPv6address>(?:\h{1,4}:){6}(?<ls32>\h{1,4}:\h{1,4}|(?<IPv4address>(?<dec-octet>[1-9]\d|1\d{2}|2[0-4]\d|25[0-5]|\d)\.\g<dec-octet>\.\g<dec-octet>\.\g<dec-octet>))|::(?:\h{1,4}:){5}\g<ls32>|\h{1,4}?::(?:\h{1,4}:){4}\g<ls32>|(?:(?:\h{1,4}:)?\h{1,4})?::(?:\h{1,4}:){3}\g<ls32>|(?:(?:\h{1,4}:){,2}\h{1,4})?::(?:\h{1,4}:){2}\g<ls32>|(?:(?:\h{1,4}:){,3}\h{1,4})?::\h{1,4}:\g<ls32>|(?:(?:\h{1,4}:){,4}\h{1,4})?::\g<ls32>|(?:(?:\h{1,4}:){,5}\h{1,4})?::\h{1,4}|(?:(?:\h{1,4}:){,6}\h{1,4})?::)|(?<IPvFuture>v\h+\.[!$&-.0-;=A-Z_a-z~]+))\])|\g<IPv4address>|(?<reg-name>(?:%\h\h|[!$&-.0-9;=A-Z_a-z~])*))(?::(?<port>\d*))?)))(?:\/|\z)/.freeze # rubocop:disable Style/RegexpLiteral, Layout/LineLength
+
             module_function
 
             def url(url, options = {})
               url!(url, options)
             rescue StandardError
-              options[:placeholder] || PLACEHOLDER
+              placeholder = options[:placeholder] || PLACEHOLDER
+
+              options[:base] == :exclude ? placeholder : "#{base_url(url)}/#{placeholder}"
             end
```

Comment:
I don't quite understand this change... Previously, if there was any error, we returned the placeholder (either the default one, or a provided one).

But now, we add the `base_url` unless `options[:base]` is `:exclude`. Which is not entirely unreasonable, but this does not seem a fix, but rather a behavior change, and a default behavior change at that.

Can you doublecheck if this is indeed what is intended?

Reply:
Glad you mention that!

> but rather a behavior change

Indeed https://github.com/DataDog/dd-trace-rb/pull/2265 changed the behaviour of `http.url` form being a path to being a proper URL.

So my reasoning was:

- `http.url` tag was path only
- `options[:placeholder]` replaced `http.url` tag on failure to parse the URI path, so 
- `http.base_url` tag was present still even on failure to parse the URI path
- there was no placeholder handling for `http.base_url`
- when `options[:base]` is `:show`, `http.url` tag goes from path to true URL
- when `options[:base]` is `:show`, `http.base_url` tag disappears and is folded into `http.url`
- To preserve the "keep url base and replace the failing path part" behaviour while having `http.url` be an actual URL, combine base with placeholder where the path would be.

Alternatives that made sense to me too are: 

- have the `options[:placeholder]` replace "whole" of `http.url` when a single part of it fails to be obtained (e.g even if we could obtain `http.base_url`), e.g `http://example.com/|` => `?`
- have the `options[:placeholder]` replace "failing parts" of `http.url`, combining it with the one we managed to have, e.g:
  - failing path: `http://example.com/|` => `http://example.com/?`
  - failing base url:
    - `http://<invalid>/path` => `?/path`
    - `<invalid>://example.com/path` => `?/path`
  - failing base url (alternative):
    - `http://<invalid>/path` => `http://?/path`
    - `<invalid>://example.com/path` => `?://example.com/path`

Note: `options[:placeholder]` does not seem to be documented, so I assumed it was some non-public configurable. Maybe it was an oversight?

WDYT?

Reply:
> Note: options[:placeholder] does not seem to be documented, so I assumed it was some non-public configurable. Maybe it was an oversight?

Sigh. I somewhat wish we were more strict with these kinds of things -- e.g. allowing customers to just provide a hash that then gets fed to an internal class directly without any filter. This means that `:placeholder` became a public config setting, even though we may not have even intended it, because you can provide it via `c.tracing.instrument :rack, quantize: ...`

I think your approach here seems reasonable. Thanks for sharing your thoughts! :)